Improve documentation of the router (#2166)
yewstack/yew: website/docs/concepts/router.md

}
```

> ### Add Parameters

Reviewer: ```suggestion
### Path Segments
```

"Parameters" can be confused with query parameters.

It is also possible to extract information from a route.

```rust
# use yew_router::prelude::*;
#[derive(Clone, Routable, PartialEq)]
enum Route {
    #[at("/")]
    Home,
    #[at("/post/:id")]
    Post { id: String },
    // ...
}
```

You can then access the post's id inside `<Switch />` and forward it to the appropriate component via properties.

> ```rust,ignore

Reviewer 2: These are not highlighting rust syntax, I wonder whats up
![image](https://user-images.githubusercontent.com/32368314/142759754-2e0f2cba-2eb3-43fa-93ac-0c6f8fe1656d.png)

Reviewer 3: Need to add a space after `rust` :) because Docusaurus' syntax highlight plugin gets confused by the trailing comma 🙃

Reviewer 2: ```suggestion
```rust ,ignore
```

fn switch(routes: &Route) -> Html {
    match routes {
        Route::Home => html! { <h1>{ "Home" }</h1> },
        Route::Post { id } => <Post {id} />,
        // ...
    }
}
```

Linking to a specific post is as easy as passing the variant to `Link`:

> ```rust,ignore

Reviewer 2: ```suggestion
```rust ,ignore
```

To navigate between pages, use either a `Link` component (which renders a `<a>` element), the `history.push` function, or the `history.replace` function, which replaces the current page in the user's browser history instead of pushing a new one onto the stack.

### Listening to Changes

In order react on route changes, you can pass a callback closure to the `listen()` method of `AnyHistory`.

> > **Note:** The history listener will get unregistered once it is dropped. Make sure to store the handle inside your component.

Reviewer: ```suggestion
:::note
The history listener will get unregistered once it is dropped. Make sure to store the handle inside your component.
:::
```
I think notes can be displayed like this

```rust
# use yew::prelude::*;

> # use yew_router::prelude::*;

Reviewer: ```suggestion
use yew_router::prelude::*;

```
Better to show the import needed.

#[function_component(Navigation)]
fn navigation() -> Html {
    let history = use_history().unwrap();

>     let handle = history.listen(|| {

Reviewer 2: This seems incorrect. Since `use_history` should re render the component with new history on changes
Or `use_location` if that does not work

Reviewer 2: `listen` method id for cases when you want updates in an non functional component

Reviewer 2: nvm scrach that, there is a `use_route` hookto get thr current route, I think we need a separate router hooks section in docs

        // handle event ...
    });

>     // save listener handle to avoid drop
>     let _listener = use_state(|| handle);

Reviewer 4: It's better to do a `use_ref ` here

```suggestion
    // use_ref will keep the listener across life cycle to avoid dropping of the listener
    use_ref(|| handle);
```

To navigate between pages, use either a `Link` component (which renders a `<a>` element), the `history.push` function, or the `history.replace` function, which replaces the current page in the user's browser history instead of pushing a new one onto the stack.

### Listening to Changes
## Functional components
Simply use available hooks `use_history`, `use_location` and `use_route`.
Your components will re-render when provided values change.

> ## Classic components

Reviewer: ```suggestion
## Struct components
```
Consistency

> In order react on route changes, you can pass a callback closure to the `listen()` method of `AnyHistory`.

Reviewer: ```suggestion
In order to react on route changes, you can pass a callback closure to the `listen()` method of `AnyHistory`.
```

To navigate between pages, use either a `Link` component (which renders a `<a>` element), the `history.push` function, or the `history.replace` function, which replaces the current page in the user's browser history instead of pushing a new one onto the stack.

> ### Listening to Changes
> 
> In order react on route changes, you can pass a callback closure to the `listen()` method of `AnyHistory`.
> 
> :::note
> The history listener will get unregistered once it is dropped. Make sure to store the handle inside your component.
> :::
> 
> ```rust
> # use yew::prelude::*;
> use yew_router::prelude::*;
> 
> #[function_component(Navigation)]
> fn navigation() -> Html {
>     let history = use_history().unwrap();
>     let handle = history.listen(|| {
>         // handle event ...
>     });
> 
>     // use_ref will keep the listener across life cycle to avoid dropping of the listener
>     use_ref(|| handle);
> 
>     html! {}
> }
> ```
> 

Reviewer 2: ```suggestion
### Listening to Changes
## Functional components
Simply use available hooks `use_history`, `use_location` and `use_route`.
Your components will re-render when provided values change.

## Classic components

In order react on route changes, you can pass a callback closure to the `listen()` method of `AnyHistory`.

:::note
The history listener will get unregistered once it is dropped. Make sure to store the handle inside your component state.
:::

```rust ,ignore
fn create(ctx: &Context<Self>) -> Self {
    let _listener = ctx.link()
        .history()
        .unwrap()
        .listen(ctx.link().callback(
            // handle event
        ));
    MyComponent {
        _listener
    }
}
```

```
